### ISSUE TITLE
How to mark or search for confirmed bugs?

### ISSUE THREAD

**Issue by AaronBallman** [llvm/llvm-project#53299]
Bugzilla had a useful feature where bugs could be marked as confirmed when someone triaged the bug. This allowed people to quickly spot bugs that someone had already validated and was a quite useful feature. Somehow, GitHub doesn't seem to have a dedicated way to do the same thing.

Can we get a new label named `confirmed` or something along those lines so that we can mark bugs after triaging them? Can we have that label applied to all bugs from Bugzilla that were in the confirmed state so we don't lose all that valuable information?

**Comment by AaronBallman** [comment]
Oh, I didn't know we could create our own labels (that sounds like we're about to be inundated with confusing one-use labels)! But yeah, I was thinking we should probably agree on how we want to triage bugs (like, is the label the "correct" way to go? I don't know.)

https://llvm.org/docs/BugLifeCycle.html#triaging-bugs is no longer really helpful and should probably be updated.

**Comment by Quuxplusone** [comment]
Right now there is a label ironically titled "new issue", which is bot-applied to every issue within a day or two of its being created — and then removed... at some point? The workflow goes "issue without labels -> issue labeled 'new issue' -> issue without labels".

I agree that
- there should be a label "confirmed";
- the label "new issue" should go away;
- the workflow should go "issue without labels -> issue labeled 'confirmed'."

I personally do have the power to script the addition of "confirmed" to all issues that were in the CONFIRMED state on Bugzilla. But I don't have the power to remove the "new issue" bot, because I don't know where it lives.

**Comment by asl** [comment]
> Right now there is a label ironically titled "new issue", which is bot-applied to every issue within a day or two of its being created

The label is created by bot immediately after the issue creation modulo usual github action queue delay if no labels were assigned by the author

> and then removed... at some point? 

Usually it is removed by whoever assigned a proper label. Sometimes - not.

Labelling of new issues was explicitly requested previously, so it should not be removed

**Comment by EugeneZelenko** [comment]
I did so only when encountered multiple issues occurrences like `clang-cl` or `compiler-rt:msan`.

I doubt that `new issue` label have too much sense. Absence of label as good as `new issue` label. And often people forget to remove `new label` after adding relevant ones or even after closing issues.

**Comment by asl** [comment]
> https://llvm.org/docs/BugLifeCycle.html#triaging-bugs is no longer really helpful and should probably be updated.

Looks like a subtask for https://github.com/llvm/llvm-project/issues/52606

**Comment by asl** [comment]
> I doubt that `new issue` label have too much sense. Absence of label as good as `new issue` label. And often people forget to remove `new label` after adding relevant ones or even after closing issues.

The same bot could listen for relabelling events and remove "new issue" label is any labels were assigned.

**Comment by Quuxplusone** [comment]
> The same bot could listen for relabelling events and remove "new issue" label if any labels were assigned.

Yes, but that would just create _two_ "noise" events in the timeline. I suggest that if #53299 remains open for a week and the original-asker-for the "new issue" bot has not stepped forward to explain why the bot is a good idea, we should just kill off the "new issue" bot, and remove the "new issue" label from all existing open and closed issues.

(Actually, [no _open_ issue has gained a "new issue" label since 5 days ago](https://github.com/llvm/llvm-project/issues?q=is%3Aissue+label%3A%22new+issue%22+is%3Aopen), so maybe the bot is already dead... but I see it applying the label to _closed_ issues as recently as 10 hours ago, so I dunno what the deal is. Anyway, I think we should assume the bot is alive until someone who owns it tells us it's dead.)

Meanwhile, regardless of the fate of "new issue", there's nothing obviously blocking us from adding the "confirmed" label — to be mechanically applied to all `everconfirmed` Bugzilla issues, and then manually applied in the traditional way from there on out. Ideally that would be done by @asl using the existing @llvmbot account, but FWIW, I could do it from my own account easily enough.

**Comment by asl** [comment]
> Yes, but that would just create _two_ "noise" events in the timeline. I suggest that if #53299 remains open for a week and the original-asker-for the "new issue" bot has not stepped forward to explain why the bot is a good idea, we should just kill off the "new issue" bot, and remove the "new issue" label from all existing open and closed issues.

Feel free to reopen https://github.com/llvm/llvm-project/issues/52614 and discuss there

> Actually, [no _open_ issue has gained a "new issue" label since 5 days ago](https://github.com/llvm/llvm-project/issues?q=is%3Aissue+label%3A%22new+issue%22+is%3Aopen), so maybe the bot is already dead...

You are not quite right. The reason for this is simple - excellent work of @EugeneZelenko and others to triage and label new issues. You can easily check by yourself any recently opened issue - there were labelling events in the timeline (e.g. https://github.com/llvm/llvm-project/issues/53295).

**Comment by AaronBallman** [comment]
> Meanwhile, regardless of the fate of "new issue", there's nothing obviously blocking us from adding the "confirmed" label — to be mechanically applied to all `everconfirmed` Bugzilla issues, and then manually applied in the traditional way from there on out. Ideally that would be done by @asl using the existing @llvmbot account, but FWIW, I could do it from my own account easily enough.

Do we think there's value to tagging any of the resolved statuses as well? (INVALID, WONTFIX, LATER, REMIND, DUPLICATE, WORKSFORME, MOVED). I'm less opinionated on that information, but some of it is quite useful (like WONTFIX clearly says we considered the bug and aren't fixing it, while WORKSFORME clearly says we can't reproduce but the bug may still be valid).

**Comment by asl** [comment]
> Do we think there's value to tagging any of the resolved statuses as well? (INVALID, WONTFIX, LATER, REMIND, DUPLICATE, WORKSFORME, MOVED). I'm less opinionated on that information, but some of it is quite useful (like WONTFIX clearly says we considered the bug and aren't fixing it, while WORKSFORME clearly says we can't reproduce but the bug may still be valid).

DUPLICATE was migrated to "duplicate" label. Note sure about the usefulness of other statuses.

**Comment by EugeneZelenko** [comment]
> Do we think there's value to tagging any of the resolved statuses as well? (INVALID, WONTFIX, LATER, REMIND, DUPLICATE, WORKSFORME, MOVED). I'm less opinionated on that information, but some of it is quite useful (like WONTFIX clearly says we considered the bug and aren't fixing it, while WORKSFORME clearly says we can't reproduce but the bug may still be valid).

There is `wontfix`, `obsolete`, `invalid`, `duplicate` labels already.

**Comment by AaronBallman** [comment]
Okay, then I think we should add `worksforme` as a label as well as it conveys useful information. I don't know that I've ever seen REMIND, LATER, or MOVED used in our Bugzilla.

**Comment by RKSimon** [comment]
I'm sorry - but why the anti "new issue" feeling? The idea was that it'd be removed later as part of triage and specific labels added.

**Comment by AaronBallman** [comment]
(FWIW, I don't have an opinion on the new issue label. I haven't found it particularly useful yet, but if others get value from it, I don't see the harm.)

**Comment by Quuxplusone** [comment]
@AaronBallman: I've added the [worksforme] label, although it remains to apply it to all relevant Bugzilla issues. I'll do that today though.

I've also consistently color-coded all the "resolution" labels (invalid, wontfix, worksforme, duplicate) as `#cccccc` (a color not used for topic labels), and documented them [on the Labels page](https://github.com/llvm/llvm-project/labels?q=Resolved).

**Comment by AaronBallman** [comment]
Thank you @Quuxplusone!

**Comment by Quuxplusone** [comment]
Update: I plan to start applying the [duplicate], [invalid], [wontfix], and [worksforme] labels about 9pm NYC time tonight. You can check my scripts here: https://github.com/Quuxplusone/BugzillaToGithub/tree/main/labelmaker
(I've just started the Step 4 script, which I expect to take three hours... er, six hours at the current rate, but I'll parallelize it... and then it's on to Step 5, which will take another two or three hours.)

**Comment by Quuxplusone** [comment]
Update: [I've updated my scripts](https://github.com/Quuxplusone/BugzillaToGithub/commit/e6cc9ef62d471e204fc9872bb14eed1a756992da) to include the final GitHub issue numbers that will be affected.
```
WORKSFORME 1290
INVALID 3126
DUPLICATE 2417
WONTFIX 1187
LATER 182
REMIND 8
MOVED 83
```
For LATER, REMIND, and MOVED: I'm not planning to mark these tonight, because I had said I wasn't. I'm ambivalent about applying them at all ever, so I guess I shouldn't. (I can't imagine anyone wanting to mark _new_ issues with these resolutions.)

For INVALID: Bugzilla had 3126 such issues; GitHub currently has 3127 issues already tagged INVALID, and spot-checking convinces me that they're the same set. So I won't mess with INVALID.

For DUPLICATE: Bugzilla had 2417 such issues; GitHub currently has 2415. I'll find out why the discrepancy, but I suspect I'll end up not messing with DUPLICATE either.

For WORKSFORME: Bugzilla had 1290 such issues; GitHub currently has 3, all new. I'll apply this label to 1290 issues.

For WONTFIX: Bugzilla had 1187 such issues; GitHub has 68, all new (or applied post-migration, e.g. bz2008/gh2380). I'll apply this label to 1187 issues.

**Comment by Quuxplusone** [comment]
Update: I've finished labeling those 18+16+1290+1187 issues with [invalid], [duplicate], [worksforme], and [wontfix].

This question remains open AFAIC: Do we have consensus to migrate over the [confirmed] label from Bugzilla's `everconfirmed` status field, and start using it to tag confirmed bugs over here? Let's see if any new voices chime in over the next few days, I guess.

**Comment by EugeneZelenko** [comment]
I think it'll be reasonable to set `confirmed` automatically only to closed issues. There are quite a lot of very old issues that may be fixed already, so it's worth to deal with them manually.

**Comment by tstellar** [comment]
If we come up with a new bug triage process or even plans to do mass issue updates with labels, I think it would be best if we presented it to the community as an RFC centered around updating: https://llvm.org/docs/BugLifeCycle.html#triaging-bugs.  
This will help make it easier for new contributors who want to triage bugs.

**Comment by AaronBallman** [comment]
Thank you for applying those labels, @Quuxplusone!

> I think it'll be reasonable to set confirmed automatically only to closed issues. There are quite a lot of very old issues that may be fixed already, so it's worth to deal with them manually.

I disagree rather strongly. Some of us (like myself) spent weeks triaging bugs only shortly before the switch to GitHub issues. Once a bug is confirmed in Bugzilla, we should continue to assume it's confirmed. It's easy enough to close bugs that have been fixed in the meantime (I also did a ton of that). This way we don't lose the work that folks have put into confirming bugs (I was especially focused on C++20 and C++23 issues).

**Comment by AaronBallman** [comment]
> If we come up with a new bug triage process or even plans to do mass issue updates with labels, I think it would be best if we presented it to the community as an RFC centered around updating: https://llvm.org/docs/BugLifeCycle.html#triaging-bugs. This will help make it easier for new contributors who want to triage bugs.

I +1'ed this because I think the community should have a documented process to follow instead of what we have now which is stale. However, I'm also not convinced we need an RFC in the first place as I don't think we need to hammer out a *new* triage process so much as we need feature parity with what we used to have in Bugzilla.

@tstellar, do you think an RFC is necessary to add the `confirmed` label? I am pretty uncomfortable with continuing to have no process to confirm bugs for an extended period of time while we (probably endlessly) debate an RFC on a new/updated triage process with no clear way to judge consensus. While someone might have other ideas for ways to triage, a `confirmed` label seems like the most direct path forward that should not be contentious as it's feature parity with what we have been using since we started tracking bugs. (This doesn't mean someone can't propose a new triage process, but that's orthogonal to having a way to mark a bug as having been independently confirmed.)

**Comment by tstellar** [comment]
@AaronBallman I don't think we need an RFC just to add a confirmed label, and use it for new bugs, especially if you are just trying to make the existing process work with GitHub, but there should at least a documentation patch that goes along with with what is being discussed here.

However, I think doing mass updates to old bugs should be limited to admins and should also be discussed on list/Discourse first.  There are some consequences to doing these updates (e.g. mass notifications or updating the modification dates) that impact the whole community.

@Quuxplusone Please hold off on any more mass updates until we have a chance to have a larger discussion about this.

**Comment by AaronBallman** [comment]
>  but there should at least a documentation patch that goes along with with what is being discussed here.

I'm happy to work on such a thing. I mostly just want to avoid going through an RFC to open up the discussion for changes to the triage process. I'll put a patch together.

**Comment by AaronBallman** [comment]
> > but there should at least a documentation patch that goes along with with what is being discussed here.
> 
> I'm happy to work on such a thing. I mostly just want to avoid going through an RFC to open up the discussion for changes to the triage process. I'll put a patch together.

I put up a strawman proposal at: https://reviews.llvm.org/D117832

**Comment by Quuxplusone** [comment]
There are only 974 bugs in the CONFIRMED state on Bugzilla. The numbers of CONFIRMED bugs per last-modified-date (Bugzilla key `delta_ts`) are:
```
{
    '2008': 2,
    '2010': 30,
    '2011': 13,
    '2012': 12,
    '2013': 4,
    '2014': 13,
    '2015': 30,
    '2016': 45,
    '2017': 41,
    '2018': 92,
    '2019': 180,
    '2020': 224,
    '2021': 288,
}
```
IMO, these numbers are low enough that we don't have to worry at all about being overwhelmed by "false positives" (bugs that were confirmed years ago but then stopped reproing and nobody noticed).

There are also 381 issues in the `REOPENED` state with `everconfirmed=1`.
There are also 16052 (!!) issues in the `NEW` state with `everconfirmed=1`. I don't understand what this signifies. https://bugs.llvm.org/show_bug.cgi?id=52549 and https://bugs.llvm.org/show_bug.cgi?id=23153 are two randomly selected examples. 
Since I don't know what `everconfirmed=1` is doing on those issues, I think it is wisest not to add _any_ labels based on the value of `everconfirmed`. But I think it's totally reasonable to add it to the 974 bugs that are actually in the CONFIRMED state right now.

Contrary to @EugeneZelenko's comment [above](https://github.com/llvm/llvm-project/issues/53299#issuecomment-1017078171), I think there's no need to add the [confirmed] label to _closed_ issues at all. They're closed; that means either they were confirmed by definition, or they've already got the [invalid] label. I wouldn't propose to touch the closed issues, because that would have a downside (time, email notifications) without an upside AFAICS.  It turns out that 48 of the CONFIRMED bugs have already been closed on GitHub, so I wouldn't touch those.

I've updated my scripts to show the 926 issues that would gain the [confirmed] label under this proposal:
https://github.com/Quuxplusone/BugzillaToGithub/blob/b6f7f49cf176365f38e7c14eadcee355bdc995e4/labelmaker/gh-map-to-github.py

**Comment by AaronBallman** [comment]
> IMO, these numbers are low enough that we don't have to worry at all about being overwhelmed by "false positives" (bugs that were confirmed years ago but then stopped reproing and nobody noticed).

+1

> There are also 381 issues in the `REOPENED` state with `everconfirmed=1`. There are also 16052 (!!) issues in the `NEW` state with `everconfirmed=1`. I don't understand what this signifies. https://bugs.llvm.org/show_bug.cgi?id=52549 and https://bugs.llvm.org/show_bug.cgi?id=23153 are two randomly selected examples. Since I don't know what `everconfirmed=1` is doing on those issues, I think it is wisest not to add _any_ labels based on the value of `everconfirmed`. But I think it's totally reasonable to add it to the 974 bugs that are actually in the CONFIRMED state right now.

Agreed. That `everconfirmed=1` behavior is very odd, I also don't see what would have marked that as having been confirmed.

> Contrary to @EugeneZelenko's comment [above](https://github.com/llvm/llvm-project/issues/53299#issuecomment-1017078171), I think there's no need to add the [confirmed] label to _closed_ issues at all. They're closed; that means either they were confirmed by definition, or they've already got the [invalid] label. I wouldn't propose to touch the closed issues, because that would have a downside (time, email notifications) without an upside AFAICS. It turns out that 48 of the CONFIRMED bugs have already been closed on GitHub, so I wouldn't touch those.

+1

**Comment by AaronBallman** [comment]
I've added the `confirmed` label to the set of labels we support, and I pushed the documentation changes in f3e22946e5c573eaa9b05d197d0c1036b21978fb. I think the only task left for this issue is applying the `confirmed` label to the migrated issues.

**Comment by asl** [comment]
The issues were tagged
